rust/lance-io/src/object_store/providers.rs

```diff
@@ -23,70 +27,278 @@ pub trait ObjectStoreProvider: std::fmt::Debug + Sync + Send {
                 location!(),
             )
         })?;
-        provider.new_store(base_path, params).await
+        let mut store = provider.new_store(base_path, params).await?;
+
+        store.inner = store.inner.traced();
+
+        if let Some(wrapper) = &params.object_store_wrapper {
+            store.inner = wrapper.wrap(store.inner);
+        }
+
+        let store = Arc::new(store);
+
+        {
+            // Insert the store into the cache
+            let mut cache_lock = self
+                .active_stores
+                .write()
+                .expect("ObjectStoreRegistry lock poisoned");
+            cache_lock.insert(cache_key, Arc::downgrade(&store));
+        }
+
+        Ok(store)
     }
 }
 
 impl Default for ObjectStoreRegistry {
     fn default() -> Self {
-        let mut registry = Self {
-            providers: HashMap::new(),
-        };
-        registry.insert("memory", Arc::new(memory::MemoryStoreProvider));
-        registry.insert("file", Arc::new(local::FileStoreProvider));
+        let mut providers: HashMap<String, Arc<dyn ObjectStoreProvider>> = HashMap::new();
+
+        providers.insert("memory".into(), Arc::new(memory::MemoryStoreProvider));
+        providers.insert("file".into(), Arc::new(local::FileStoreProvider));
         // The "file" scheme has special optimized code paths that bypass
         // the ObjectStore API for better performance. However, this can make it
         // hard to test when using ObjectStore wrappers, such as IOTrackingStore.
         // So we provide a "file-object-store" scheme that uses the ObjectStore API.
         // The specialized code paths are differentiated by the scheme name.
-        registry.insert("file-object-store", Arc::new(local::FileStoreProvider));
+        providers.insert(
+            "file-object-store".into(),
+            Arc::new(local::FileStoreProvider),
+        );
+
         #[cfg(feature = "aws")]
         {
             let aws = Arc::new(aws::AwsStoreProvider);
-            registry.insert("s3", aws.clone());
-            registry.insert("s3+ddb", aws);
+            providers.insert("s3".into(), aws.clone());
+            providers.insert("s3+ddb".into(), aws);
         }
         #[cfg(feature = "azure")]
-        registry.insert("az", Arc::new(azure::AzureBlobStoreProvider));
+        providers.insert("az".into(), Arc::new(azure::AzureBlobStoreProvider));
         #[cfg(feature = "gcp")]
-        registry.insert("gs", Arc::new(gcp::GcsStoreProvider));
-        registry
+        providers.insert("gs".into(), Arc::new(gcp::GcsStoreProvider));
+        Self {
+            providers: RwLock::new(providers),
+            active_stores: RwLock::new(HashMap::new()),
+        }
     }
 }
 
 impl ObjectStoreRegistry {
-    pub fn insert(&mut self, scheme: &str, provider: Arc<dyn ObjectStoreProvider>) {
-        self.providers.insert(scheme.into(), provider);
+    /// Add a new object store provider to the registry. This will be called
```

Comment:
```suggestion
    /// Add a new object store provider to the registry. The provider will be used
```

Minor wording as I read it as "this method (insert) will be called..."